File: go/mechanisms/svm/swig.go
package svm

import (
	"encoding/binary"
	"errors"
	"fmt"
	"sort"

	solana "github.com/gagliardetto/solana-go"
)

// SwigNormalizer detects and flattens Swig smart-wallet transactions into the
// same NormalizedTransaction shape used by regular transactions.
type SwigNormalizer struct{}

func (s *SwigNormalizer) CanHandle(tx *solana.Transaction) bool {
	return IsSwigTransaction(tx)
}

func (s *SwigNormalizer) Normalize(tx *solana.Transaction, ctx *NormalizationContext) (*NormalizedTransaction, error) {
	result, err := ParseSwigTransaction(tx)
	if err != nil {
		return nil, err
	}
	instructions := result.Instructions
	if ctx != nil {
		instructions, err = FilterFeeTransfers(tx, instructions, ctx.Asset, ctx.PayTo, ctx.SignerAddresses)
		if err != nil {
			return nil, err
		}
	}
	return &NormalizedTransaction{
		Instructions: instructions,
		Payer:        result.SwigPDA,
	}, nil
}

// SwigCompactInstruction is a decoded compact instruction embedded in a Swig
// SignV2 instruction payload.  Indices reference the SignV2 instruction's
// own account list, not the outer transaction's global account keys.
type SwigCompactInstruction struct {
	ProgramIDIndex uint8
	Accounts       []uint8
	Data           []byte
}

// DecodeSwigCompactInstructions parses the compact instructions embedded in the
// data payload of a Swig SignV2 instruction.
//
// Outer instruction data layout:
//
//	[0..1]  discriminator         U16 LE
//	[2..3]  instructionPayloadLen U16 LE
//	[4..7]  roleId                U32 LE
//	[8..]   compact instructions  (instructionPayloadLen bytes)
//
// Compact instructions payload:
//
//	[0]         numInstructions U8
//	[1..]       compact instruction entries...
//
// Each CompactInstruction:
//
//	[0]         programIDIndex U8
//	[1]         numAccounts    U8
//	[2..N+1]    accounts       []U8
//	[N+2..N+3]  dataLen        U16 LE
//	[N+4..]     data           raw bytes
func DecodeSwigCompactInstructions(data []byte) ([]SwigCompactInstruction, error) {
	if len(data) < 4 {
		return nil, fmt.Errorf("swig instruction data too short: need ≥4 bytes, got %d", len(data))
	}

	instructionPayloadLen := int(binary.LittleEndian.Uint16(data[2:4]))
	startOffset := 8
	if len(data) < startOffset+instructionPayloadLen {
		return nil, fmt.Errorf("swig instruction data truncated: payload needs %d bytes but only %d available after offset %d",
			instructionPayloadLen, len(data)-startOffset, startOffset)
	}

	var results []SwigCompactInstruction
	offset := startOffset + 1 // skip numInstructions count byte
	endOffset := startOffset + instructionPayloadLen

	for offset < endOffset {
		if offset >= len(data) {
			break
		}

		// programIDIndex: U8
		programIDIndex := data[offset]
		offset++

		// numAccounts: U8
		if offset >= endOffset {
			break
		}
		numAccounts := int(data[offset])
		offset++

		// accounts: []U8
		if offset+numAccounts > endOffset {
			break
		}
		accounts := make([]uint8, numAccounts)
		copy(accounts, data[offset:offset+numAccounts])
		offset += numAccounts

		// dataLen: U16 LE
		if offset+2 > endOffset {
			break
		}
		dataLen := int(binary.LittleEndian.Uint16(data[offset : offset+2]))
		offset += 2

		// instruction data
		if offset+dataLen > endOffset {
			break
		}
		instrData := make([]byte, dataLen)
		copy(instrData, data[offset:offset+dataLen])
		offset += dataLen

		results = append(results, SwigCompactInstruction{
			ProgramIDIndex: programIDIndex,
			Accounts:       accounts,
			Data:           instrData,
		})
	}

	return results, nil
}

// IsSwigTransaction returns true when the transaction has a Swig layout:
//   - Every instruction is one of: compute budget, secp256r1 precompile, or Swig SignV2
//   - At least one Swig SignV2 instruction is present
func IsSwigTransaction(tx *solana.Transaction) bool {
	instructions := tx.Message.Instructions
	if len(instructions) == 0 {
		return false
	}

	secp256r1Pubkey := solana.MustPublicKeyFromBase58(Secp256r1PrecompileAddress)
	swigPubkey := solana.MustPublicKeyFromBase58(SwigProgramAddress)

	hasSignV2 := false
	for _, inst := range instructions {
		if int(inst.ProgramIDIndex) >= len(tx.Message.AccountKeys) {
			return false
		}
		progID := tx.Message.AccountKeys[inst.ProgramIDIndex]
		if progID.Equals(solana.ComputeBudget) || progID.Equals(secp256r1Pubkey) {
			continue
		}
		if progID.Equals(swigPubkey) {
			if len(inst.Data) < 2 {
				return false
			}
			discriminator := binary.LittleEndian.Uint16(inst.Data[0:2])
			if discriminator != SwigSignV2Discriminator {
				return false
			}
			hasSignV2 = true
			continue
		}
		return false // unrecognized instruction
	}
	return hasSignV2
}

// ParseSwigResult holds the flattened instructions and the Swig PDA address.
type ParseSwigResult struct {
	Instructions []solana.CompiledInstruction
	SwigPDA      string
}

// ParseSwigTransaction flattens a Swig transaction into the same instruction
// layout as a regular one. It collects non-precompile, non-SignV2 outer
// instructions (compute budgets) and resolves the compact instructions embedded
// in each SignV2 instruction.
//
// A transaction may contain multiple SignV2 instructions. All must reference
// the same Swig PDA (accounts[0]).
func ParseSwigTransaction(tx *solana.Transaction) (*ParseSwigResult, error) {
	instructions := tx.Message.Instructions
	if len(instructions) == 0 {
		return nil, errors.New("no instructions")
	}

	secp256r1Pubkey := solana.MustPublicKeyFromBase58(Secp256r1PrecompileAddress)
	swigPubkey := solana.MustPublicKeyFromBase58(SwigProgramAddress)

	// 1. Single pass: separate SignV2 instructions from the rest
	var result []solana.CompiledInstruction
	var signV2Instructions []solana.CompiledInstruction
	for _, inst := range instructions {
		progID := tx.Message.AccountKeys[inst.ProgramIDIndex]
		if progID.Equals(secp256r1Pubkey) {
			continue // skip precompile
		}
		if progID.Equals(swigPubkey) {
			signV2Instructions = append(signV2Instructions, inst)
		} else {
			result = append(result, inst) // compute budget and other non-precompile instructions
		}
	}

	// Sort compute budget instructions so SetComputeUnitLimit (disc=2) precedes
	// SetComputeUnitPrice (disc=3), matching the order the facilitator expects.
	sort.Slice(result, func(i, j int) bool {
		iIsComputeBudget := tx.Message.AccountKeys[result[i].ProgramIDIndex].Equals(solana.ComputeBudget)
		jIsComputeBudget := tx.Message.AccountKeys[result[j].ProgramIDIndex].Equals(solana.ComputeBudget)
		if iIsComputeBudget && jIsComputeBudget && len(result[i].Data) > 0 && len(result[j].Data) > 0 {
			return result[i].Data[0] < result[j].Data[0]
		}
		return false
	})

	if len(signV2Instructions) == 0 {
		return nil, errors.New("invalid_exact_svm_payload_no_transfer_instruction")
	}

	// 2. Process each SignV2 instruction
	swigPDA := ""
	for _, signV2 := range signV2Instructions {
		// Extract Swig PDA from SignV2's first account
		if len(signV2.Accounts) < 2 {
			return nil, errors.New("invalid_exact_svm_payload_no_transfer_instruction")
		}
		if int(signV2.Accounts[0]) >= len(tx.Message.AccountKeys) {
			return nil, errors.New("invalid_exact_svm_payload_no_transfer_instruction")
		}
		swigConfigKey := tx.Message.AccountKeys[signV2.Accounts[0]]
		pda := swigConfigKey.String()

		// Enforce all SignV2 instructions share the same PDA
		if swigPDA == "" {
			swigPDA = pda
		} else if pda != swigPDA {
			return nil, errors.New("swig_pda_mismatch: all SignV2 instructions must reference the same Swig PDA")
		}

		// Validate Swig wallet address derivation (cross-check accounts[0] and accounts[1])
		if int(signV2.Accounts[1]) >= len(tx.Message.AccountKeys) {
			return nil, errors.New("invalid_swig_wallet_address_derivation")
		}
		actualWalletAddress := tx.Message.AccountKeys[signV2.Accounts[1]]
		expectedWalletAddress, _, err := solana.FindProgramAddress(
			[][]byte{[]byte("swig-wallet-address"), swigConfigKey.Bytes()},
			swigPubkey,
		)
		if err != nil {
			return nil, fmt.Errorf("failed to derive swig wallet address: %w", err)
		}
		if !actualWalletAddress.Equals(expectedWalletAddress) {
			return nil, errors.New("invalid_swig_wallet_address_derivation")
		}

		// Decode compact instructions from SignV2 data
		compactInstructions, err := DecodeSwigCompactInstructions(signV2.Data)
		if err != nil {
			return nil, err
		}

		// Resolve compact instructions: remap local indices through signV2.Accounts
		for _, ci := range compactInstructions {
			if int(ci.ProgramIDIndex) >= len(signV2.Accounts) {
				return nil, fmt.Errorf("compact instruction programIDIndex %d out of range (signV2 has %d accounts)",
					ci.ProgramIDIndex, len(signV2.Accounts))
			}
			accounts := make([]uint16, len(ci.Accounts))
			for j, a := range ci.Accounts {
				if int(a) >= len(signV2.Accounts) {
					return nil, fmt.Errorf("compact instruction account index %d out of range (signV2 has %d accounts)",
						a, len(signV2.Accounts))
				}
				accounts[j] = signV2.Accounts[a]
			}
			result = append(result, solana.CompiledInstruction{
				ProgramIDIndex: signV2.Accounts[ci.ProgramIDIndex],
				Accounts:       accounts,
				Data:           ci.Data,
			})
		}
	}

	return &ParseSwigResult{
		Instructions: result,
		SwigPDA:      swigPDA,
	}, nil
}

// FilterFeeTransfers removes Swig fee transfers from a flattened instruction list,
// returning [ComputeLimit, ComputePrice, MerchantTransfer, ...nonTransferTail].
//
// It identifies the merchant transfer by matching the destination ATA derived from
// asset + payTo. All other TransferChecked instructions are considered fee transfers
// and are stripped, after a safety check that they don't touch signer or payTo addresses.
func FilterFeeTransfers(
	tx *solana.Transaction,
	instructions []solana.CompiledInstruction,
	asset string,
	payTo string,
	signerAddresses []string,
) ([]solana.CompiledInstruction, error) {
	// Derive expected merchant ATA
	payToPubkey, err := solana.PublicKeyFromBase58(payTo)
	if err != nil {
		return nil, fmt.Errorf("invalid payTo address: %w", err)
	}
	mintPubkey, err := solana.PublicKeyFromBase58(asset)
	if err != nil {
		return nil, fmt.Errorf("invalid asset address: %w", err)
	}
	expectedMerchantATA, _, err := solana.FindAssociatedTokenAddress(payToPubkey, mintPubkey)
	if err != nil {
		return nil, fmt.Errorf("failed to derive merchant ATA: %w", err)
	}

	var merchantTransfer *solana.CompiledInstruction
	var nonTransferTail []solana.CompiledInstruction

	for i := 2; i < len(instructions); i++ {
		inst := instructions[i]

		// Try to decode as a TransferChecked instruction
		isTC, sourceAddr, destAddr := tryDecodeTransferChecked(tx, inst)
		if !isTC {
			// Not a TransferChecked — keep as non-transfer tail
			nonTransferTail = append(nonTransferTail, inst)
			continue
		}

		if destAddr == expectedMerchantATA.String() {
			merchantTransfer = &instructions[i]
		} else {
			// Fee transfer — safety check: neither source nor destination touches signer or payTo
			for _, signerAddr := range signerAddresses {
				if sourceAddr == signerAddr || destAddr == signerAddr {
					return nil, errors.New("invalid_exact_solana_payload_suspicious_fee_transfer")
				}
			}
			if sourceAddr == payTo || destAddr == payTo {
				return nil, errors.New("invalid_exact_solana_payload_suspicious_fee_transfer")
			}
			// Fee transfer is safe — strip it from output
		}
	}

	if merchantTransfer == nil {
		return nil, errors.New("invalid_exact_svm_payload_no_transfer_instruction")
	}

	result := []solana.CompiledInstruction{instructions[0], instructions[1], *merchantTransfer}
	result = append(result, nonTransferTail...)
	return result, nil
}

// tryDecodeTransferChecked checks if a compiled instruction is a SPL TransferChecked
// by inspecting the program ID and data discriminator. Returns whether it's a
// TransferChecked plus the source and destination addresses.
//
// This avoids ResolveInstructionAccounts which doesn't work with ALT-resolved
// messages. Instead it indexes directly into tx.Message.AccountKeys.
func tryDecodeTransferChecked(tx *solana.Transaction, inst solana.CompiledInstruction) (bool, string, string) {
	accountKeys := tx.Message.AccountKeys
	if int(inst.ProgramIDIndex) >= len(accountKeys) {
		return false, "", ""
	}
	progID := accountKeys[inst.ProgramIDIndex]
	if progID != solana.TokenProgramID && progID != solana.Token2022ProgramID {
		return false, "", ""
	}

	// TransferChecked discriminator = 12, minimum data = 10 bytes (1 disc + 8 amount + 1 decimals)
	if len(inst.Data) < 10 || inst.Data[0] != 12 {
		return false, "", ""
	}

	// TransferChecked accounts: [source(0), mint(1), destination(2), authority(3)]
	if len(inst.Accounts) < 4 {
		return false, "", ""
	}

	sourceIdx := int(inst.Accounts[0])
	destIdx := int(inst.Accounts[2])
	if sourceIdx >= len(accountKeys) || destIdx >= len(accountKeys) {
		return false, "", ""
	}

	return true, accountKeys[sourceIdx].String(), accountKeys[destIdx].String()
}

File: python/x402/mechanisms/svm/swig.py
"""Swig smart wallet transaction detection, parsing, and flattening."""

from dataclasses import dataclass

try:
    from solders.pubkey import Pubkey
    from solders.transaction import VersionedTransaction
except ImportError as e:
    raise ImportError(
        "SVM mechanism requires solana packages. Install with: pip install x402[svm]"
    ) from e

from .constants import (
    COMPUTE_BUDGET_PROGRAM_ADDRESS,
    ERR_NO_TRANSFER_INSTRUCTION,
    ERR_SUSPICIOUS_FEE_TRANSFER,
    SECP256R1_PRECOMPILE_ADDRESS,
    SWIG_PROGRAM_ADDRESS,
    SWIG_SIGN_V2_DISCRIMINATOR,
    TOKEN_2022_PROGRAM_ADDRESS,
    TOKEN_PROGRAM_ADDRESS,
)
from .utils import derive_ata


@dataclass
class SwigCompactInstruction:
    """A decoded compact instruction embedded in a Swig SignV2 payload.

    Indices reference the SignV2 instruction's own account list,
    not the outer transaction's global account keys.
    """

    program_id_index: int
    accounts: list[int]
    data: bytes


@dataclass
class NormalizedInstruction:
    """An instruction with global indices into tx.message.account_keys.

    Mimics the solders CompiledInstruction interface so the facilitator
    can use it interchangeably.
    """

    program_id_index: int
    accounts: list[int]
    data: bytes


@dataclass
class ParseSwigResult:
    """Result of parsing/flattening a Swig transaction."""

    instructions: list[NormalizedInstruction]
    swig_pda: str


def is_swig_transaction(tx: VersionedTransaction) -> bool:
    """Check if a transaction has a Swig smart wallet layout.

    A Swig transaction contains only compute budget, secp256r1 precompile,
    and Swig SignV2 instructions, with at least one SignV2 present.
    """
    instructions = tx.message.instructions
    if len(instructions) == 0:
        return False

    account_keys = list(tx.message.account_keys)
    compute_budget = Pubkey.from_string(COMPUTE_BUDGET_PROGRAM_ADDRESS)
    secp256r1 = Pubkey.from_string(SECP256R1_PRECOMPILE_ADDRESS)
    swig = Pubkey.from_string(SWIG_PROGRAM_ADDRESS)

    has_sign_v2 = False
    for ix in instructions:
        if ix.program_id_index >= len(account_keys):
            return False
        prog_id = account_keys[ix.program_id_index]
        if prog_id == compute_budget or prog_id == secp256r1:
            continue
        if prog_id == swig:
            ix_data = bytes(ix.data)
            if len(ix_data) < 2:
                return False
            discriminator = int.from_bytes(ix_data[0:2], "little")
            if discriminator != SWIG_SIGN_V2_DISCRIMINATOR:
                return False
            has_sign_v2 = True
            continue
        return False  # unrecognized instruction

    return has_sign_v2


def decode_swig_compact_instructions(data: bytes) -> list[SwigCompactInstruction]:
    """Decode compact instructions from a Swig SignV2 instruction payload.

    Data layout:
        [0..1]  discriminator         U16 LE
        [2..3]  instructionPayloadLen U16 LE
        [4..7]  roleId                U32 LE
        [8..]   compact instructions  (instructionPayloadLen bytes)

    Compact instructions payload:
        [0]         numInstructions U8
        [1..]       compact instruction entries...

    Each compact instruction:
        [0]         programIDIndex U8
        [1]         numAccounts    U8
        [2..N+1]    accounts       []U8
        [N+2..N+3]  dataLen        U16 LE
        [N+4..]     data           raw bytes
    """
    if len(data) < 4:
        raise ValueError(
            f"swig instruction data too short: need \u22654 bytes, got {len(data)}"
        )

    instruction_payload_len = int.from_bytes(data[2:4], "little")
    start_offset = 8
    if len(data) < start_offset + instruction_payload_len:
        raise ValueError(
            f"swig instruction data truncated: payload needs {instruction_payload_len} bytes "
            f"but only {len(data) - start_offset} available after offset {start_offset}"
        )

    results: list[SwigCompactInstruction] = []
    offset = start_offset + 1  # skip numInstructions count byte
    end_offset = start_offset + instruction_payload_len

    while offset < end_offset:
        if offset >= len(data):
            break

        # programIDIndex: U8
        program_id_index = data[offset]
        offset += 1

        # numAccounts: U8
        if offset >= end_offset:
            break
        num_accounts = data[offset]
        offset += 1

        # accounts: []U8
        if offset + num_accounts > end_offset:
            break
        accounts = list(data[offset : offset + num_accounts])
        offset += num_accounts

        # dataLen: U16 LE
        if offset + 2 > end_offset:
            break
        data_len = int.from_bytes(data[offset : offset + 2], "little")
        offset += 2

        # instruction data
        if offset + data_len > end_offset:
            break
        instr_data = bytes(data[offset : offset + data_len])
        offset += data_len

        results.append(
            SwigCompactInstruction(
                program_id_index=program_id_index,
                accounts=accounts,
                data=instr_data,
            )
        )

    return results


def parse_swig_transaction(tx: VersionedTransaction) -> ParseSwigResult:
    """Flatten a Swig transaction into a regular instruction layout.

    Collects non-precompile, non-SignV2 outer instructions (compute budgets)
    and resolves the compact instructions embedded in each SignV2 instruction.

    All SignV2 instructions must reference the same Swig PDA (accounts[0]).
    """
    instructions = tx.message.instructions
    if len(instructions) == 0:
        raise ValueError("no instructions")

    account_keys = list(tx.message.account_keys)
    secp256r1 = Pubkey.from_string(SECP256R1_PRECOMPILE_ADDRESS)
    swig_pubkey = Pubkey.from_string(SWIG_PROGRAM_ADDRESS)

    # 1. Single pass: separate SignV2 from the rest
    result: list[NormalizedInstruction] = []
    sign_v2_instructions = []
    for ix in instructions:
        prog_id = account_keys[ix.program_id_index]
        if prog_id == secp256r1:
            continue  # skip precompile
        if prog_id == swig_pubkey:
            sign_v2_instructions.append(ix)
        else:
            # compute budget and other non-precompile instructions (pass through)
            result.append(
                NormalizedInstruction(
                    program_id_index=ix.program_id_index,
                    accounts=list(ix.accounts),
                    data=bytes(ix.data),
                )
            )

    if len(sign_v2_instructions) == 0:
        raise ValueError("invalid_exact_svm_payload_no_transfer_instruction")

    # Sort compute budget instructions so SetComputeUnitLimit (disc=2) precedes
    # SetComputeUnitPrice (disc=3), matching the order the facilitator expects.
    compute_budget = Pubkey.from_string(COMPUTE_BUDGET_PROGRAM_ADDRESS)
    result.sort(
        key=lambda ix: ix.data[0]
        if account_keys[ix.program_id_index] == compute_budget and len(ix.data) > 0
        else 0
    )

    # 2. Process each SignV2 instruction
    swig_pda = ""
    for sign_v2 in sign_v2_instructions:
        sign_v2_accounts = list(sign_v2.accounts)

        # Extract Swig PDA from SignV2's first account
        if len(sign_v2_accounts) < 2:
            raise ValueError("invalid_exact_svm_payload_no_transfer_instruction")
        if sign_v2_accounts[0] >= len(account_keys):
            raise ValueError("invalid_exact_svm_payload_no_transfer_instruction")

        swig_config_key = account_keys[sign_v2_accounts[0]]
        pda = str(swig_config_key)

        # Enforce all SignV2 instructions share the same PDA
        if swig_pda == "":
            swig_pda = pda
        elif pda != swig_pda:
            raise ValueError(
                "swig_pda_mismatch: all SignV2 instructions must reference the same Swig PDA"
            )

        # Validate Swig wallet address derivation (cross-check accounts[0] and accounts[1])
        if sign_v2_accounts[1] >= len(account_keys):
            raise ValueError("invalid_swig_wallet_address_derivation")
        actual_wallet_address = account_keys[sign_v2_accounts[1]]
        expected_wallet_address, _ = Pubkey.find_program_address(
            [b"swig-wallet-address", bytes(swig_config_key)],
            swig_pubkey,
        )
        if actual_wallet_address != expected_wallet_address:
            raise ValueError("invalid_swig_wallet_address_derivation")

        # Decode compact instructions from SignV2 data
        compact_instructions = decode_swig_compact_instructions(bytes(sign_v2.data))

        # Resolve compact instructions: remap local indices through sign_v2.accounts
        for ci in compact_instructions:
            if ci.program_id_index >= len(sign_v2_accounts):
                raise ValueError(
                    f"compact instruction programIDIndex {ci.program_id_index} "
                    f"out of range (signV2 has {len(sign_v2_accounts)} accounts)"
                )
            remapped_accounts = []
            for a in ci.accounts:
                if a >= len(sign_v2_accounts):
                    raise ValueError(
                        f"compact instruction account index {a} "
                        f"out of range (signV2 has {len(sign_v2_accounts)} accounts)"
                    )
                remapped_accounts.append(sign_v2_accounts[a])
            result.append(
                NormalizedInstruction(
                    program_id_index=sign_v2_accounts[ci.program_id_index],
                    accounts=remapped_accounts,
                    data=ci.data,
                )
            )

    return ParseSwigResult(instructions=result, swig_pda=swig_pda)


def _try_decode_transfer_checked(
    account_keys: list[Pubkey], inst: NormalizedInstruction
) -> tuple[bool, str, str]:
    """Check if a normalized instruction is SPL TransferChecked.

    Returns (is_transfer_checked, source_address, dest_address).
    """
    if inst.program_id_index >= len(account_keys):
        return False, "", ""

    prog_id = account_keys[inst.program_id_index]
    token_program = Pubkey.from_string(TOKEN_PROGRAM_ADDRESS)
    token_2022_program = Pubkey.from_string(TOKEN_2022_PROGRAM_ADDRESS)
    if prog_id != token_program and prog_id != token_2022_program:
        return False, "", ""

    if len(inst.data) < 10 or inst.data[0] != 12:
        return False, "", ""

    if len(inst.accounts) < 4:
        return False, "", ""

    source_idx = inst.accounts[0]
    dest_idx = inst.accounts[2]
    if source_idx >= len(account_keys) or dest_idx >= len(account_keys):
        return False, "", ""

    return True, str(account_keys[source_idx]), str(account_keys[dest_idx])


def filter_fee_transfers(
    account_keys: list[Pubkey],
    instructions: list[NormalizedInstruction],
    asset: str,
    pay_to: str,
    signer_addresses: list[str],
) -> list[NormalizedInstruction]:
    """Remove Swig fee transfers, keeping only the merchant transfer.

    Returns [ComputeLimit, ComputePrice, MerchantTransfer, ...nonTransferTail].
    """
    expected_merchant_ata = derive_ata(pay_to, asset, TOKEN_PROGRAM_ADDRESS)
    expected_merchant_ata_2022 = derive_ata(pay_to, asset, TOKEN_2022_PROGRAM_ADDRESS)

    merchant_transfer: NormalizedInstruction | None = None
    non_transfer_tail: list[NormalizedInstruction] = []

    for i in range(2, len(instructions)):
        inst = instructions[i]
        is_tc, source_addr, dest_addr = _try_decode_transfer_checked(account_keys, inst)

        if not is_tc:
            non_transfer_tail.append(inst)
            continue

        if dest_addr == expected_merchant_ata or dest_addr == expected_merchant_ata_2022:
            merchant_transfer = inst
        else:
            # Fee transfer — safety check
            for signer_addr in signer_addresses:
                if source_addr == signer_addr or dest_addr == signer_addr:
                    raise ValueError(ERR_SUSPICIOUS_FEE_TRANSFER)
            if source_addr == pay_to or dest_addr == pay_to:
                raise ValueError(ERR_SUSPICIOUS_FEE_TRANSFER)

    if merchant_transfer is None:
        raise ValueError(ERR_NO_TRANSFER_INSTRUCTION)

    return [instructions[0], instructions[1], merchant_transfer, *non_transfer_tail]
